### ISSUE TITLE
Creating  mgbf interface in saber/jedi

### ISSUE THREAD

**Issue by TingLei-NOAA** [NOAA-EMC/RDASApp#31]
As a part of the EMC efforts for MGBF (Multigrid Beta function modeling of background error covariances) for regional JEDI,  a MGBF interface in saber/jedi using the newly developed oops interfaced MGBF lib is being created, which is supposed to be used as starting point/testbed for MGBF developers directly to work on JEDI applications. 
This issue will facilitate communications and collaborations for MGBF with JEDI. 
Relevant important design decisions by MGBF  developers at EMC include:    the regular grids will be used for the filtering grids and the mapping between these filtering grids and the analysis grids (native model grids in jedi) will be deal with by a rather generic atlas based JEDI interpolator  which is actively being developed by JEDI core team.  
It is noted: the GSIBE component in saber has been used as  a good example for this work.

**Comment by TingLei-NOAA** [comment]
An update on May 1,2024:

1) It was planned to create a mgbf owned interpolation class as the subclass of the central block to do the interpolation (and its adjoint) to provide the mapping between the mgbf filtering grids and the analysis grids.  However, it is found the saber::interpolation  is being developed, the interface of which has been in saber, to be available maybe in several weeks to be tested/made use of  (Thanks to F. Herbert at JCSDA) . Hence, our mgbf interpolation part is given up and  a  thus-simplified mgbf-jedi interface was done with a few imiportant details to be implemented later, which including the passing of data between jedi fieldset and mgbf "native grid". This version is at the private branch feature/mgbf_tl 
2) A important problem to be addressed is that the atlas support to regional domain is found still to be developed . see https://github.com/JCSDA-internal/oops/issues/2575#issuecomment-2085124575 and https://github.com/ecmwf/atlas/discussions/190.  Further exploring on this is ongoing and  how to proceed in collaboration with jedi core teams is to be discussed and decided.

**Comment by TingLei-NOAA** [comment]
Thanks to suggestions from @danholdaway , it is decided to use Saber::gsi::interpolation, which has been working for conversion between unstructured grids and lat/lon grids on global domains and Dan believe it should also work for regional domains.  Hence it is a good tool for what current mgbf needs.

**Comment by danholdaway** [comment]
@TingLei-NOAA as I mentioned, the biggest thing we now need to ensure is that the domain of the model is completely contained by the domain of the background error model. We should put our heads together to figure out a generic way to do that. It should be a check in Saber somewhere.

**Comment by TingLei-NOAA** [comment]
@danholdaway  Sure. That needs to be taken care of when the "mgbf grids"  is created, which is now to be planned to create on fly.  There are some tricky things I need to sort out when the layout of mpi ranks are pre-defined.  I will keep discussions among us  ,maybe, in various ways.

**Comment by TingLei-NOAA** [comment]
A mgbf in jedi/saber using unstructured interpolator from saber::gsi had been created with successful compiling. 
A key choice is the definition of the type of the mgbf grids. Now it is defined as 
```
  const atlas::StructuredGrid grid(conf);
//clt  how about PointCloud functionspace
   const atlas::functionspace::StructuredColumns mgbfGridFuncSpace_(grid);
   
  ```
 Sanity tests are being setup.  The following coding/debugging through various tests  would delve into  "under-the-hood" workings of  Atlas and the unstructured interpolator on regional domains

**Comment by TingLei-NOAA** [comment]
Previously, the structuredcolumn functionspace was used as : 
```
  mgbfFuncSpace_=atlas::functionspace::StructuredColumns(grid)
 ```
 However, it is found that in Structuredcolumns functionspace, there is no lat and lon information about grid points. 
The lonlat method returned the x/y fields as in atlas/functionspace/detail/StructuredColumns.h 
```
Field lonlat() const override { return field_xy_; }
```
Hence, now PointCloud functionspace is used and being tested.

**Comment by TingLei-NOAA** [comment]
Using the pointcloud of atlas::functionspace, the mgbf Dirac test  produced a reasonable result as shown below.
![image](https://github.com/NOAA-EMC/RDASApp/assets/63461756/83d2245e-e084-41e4-bda7-192c39f9475d)
Further verification (including tests for mpas) and development will be carried out  on top of this working version of mgbf in jedi.
This working version also confirms both the efficiency and validity of the mgbf implementation strategy designed and decided by the team, including the choice of the gsi::unstructured_interpolator according suggested by  @danholdaway

**Comment by JimPurser-NOAA** [comment]
Hi Ting:

What is the blotch that we are looking at? Can you explain the context
please?

Jim

**Comment by TingLei-NOAA** [comment]
@JimPurser-NOAA  In this Dirac test,   the variable (No2 here) (on the model grids)  is assigned 1 on one point and zero on all other points. Then this field (vector) was multiplied by the B matrix . The above figure shows the field resulting from such operation. 
For mgbf, the B is composed of MGBF on a regular grid as the central block and the interpolation (from the former regular grids to model grids) and its adjoint on both sides.    Hope this explanation helps.

**Comment by JimPurser-NOAA** [comment]
Hi Ting:

Is this a univariate analysis of a single scalar quantity, then?

If so, why is the response so anisotropic -- it just looks like a squashed
bug!

Thanks,
Jim

**Comment by TingLei-NOAA** [comment]
Jim,  Yes. in this dirac test, it is a just a univariate "operation".   I will keep your points on mind in the future more strict verification/validation. For being now, the factors i think of could contribute to the "unstructured" /un-smoothed "anistropic patter include :1) the resolution of the mgbf filtering grids;  2) the plot from ncview , in which ncview treat the points as on regular grids while the fv3 grids are not .  3) possible setup for the mgbf in my test.    With the problems you raised , which are to be answered  decisively in the future,  I think the results are satisfying to me to show the mgbf in saber/jedi is working and hopefully, you and other colleagues would agree on this :).

**Comment by TingLei-NOAA** [comment]
Following discussion with Jim,  another dirac test using 100km resolution (compared with previous 200km resolution) of the filtering grids, the result is shown below:
  
![image](https://github.com/NOAA-EMC/RDASApp/assets/63461756/f38b264b-aac1-4b56-856c-88eddbc780f2)
It seems the patter became smoother but still some irregularities are obvious. 
Next , we will see the impacts of correlation length scales modelled in mgbf.

**Comment by TingLei-NOAA** [comment]
The verification/validation/debugging is ongoing by tuning of parameters and tweaking codes. 
The mgbf interpolation(and its adjoint) part (using saber::gsi::unstructor_interplator) have passed this cycle of verification and debugging, which are run as outer saber chain blocks.  
The self adjoint test failed (invoked by saber's generic adjoint test interface), but it failed for the created random vector on the model grids contain nan values and is estimated not relevant to the mgbf coding, though it would be revisited in the future. 
A series of tests for tests of the mgbf interpolation have been conduced, which collectively verified the former.  please contact me for more details.  Below, I would only show one test, in which, the mgbf processing was skipped and the dirac test is infact compose of the interpolation (from model grids to mgbf filtering grids) following its adjoint initiated by the dirac vector. 
Below is first the domains for mgbf filtering grids and model (As @danholdaway stressed, the former should be larger ).
![image](https://github.com/NOAA-EMC/RDASApp/assets/63461756/4719e9df-bb6b-4e85-869d-54d41a03e086)
Then the increment of such "dirac" test 
![image](https://github.com/NOAA-EMC/RDASApp/assets/63461756/b8a5bbd9-c288-4d3b-82b0-1ea9c6e3a396)

Hence, for this cycle, the following verification/debugging would focus on the mgbf step (the central chain block)

**Comment by TingLei-NOAA** [comment]
With @MiodragRancic-NOAA 's help, the appropriate procedure in mgbf lib has been selected  and now, the mgbf at JEDI apparrently works "similar" to the established mgbf behaviors as shown in the attached dirac tests
![image](https://github.com/user-attachments/assets/579338c6-3ed7-4fab-bd45-865a13321ba7)

Further verifications and the calibration of the mgbf coefficients are ongoing

**Comment by TingLei-NOAA** [comment]
A fv3jedi 3DVar run using mgbf background function has been completed with a single NO2 observation. The results are reasonable and further  investigation, including comparison with bump, are to be done.
![image](https://github.com/user-attachments/assets/72446173-1652-43e0-ac81-959526970f0b)
The tests for regional mpas is also being worked on

**Comment by TingLei-NOAA** [comment]
A mgbf dirac test using two analysis variables had been done showing reasonable results after fixing a bug in the variable passing from jedi fields. (see update july 21,2024 at https://docs.google.com/presentation/d/1Xu01QHNSgZTHrT-BaVVTEuDk8ENbZsed/edit#slide=id.g2ed53956188_0_0) 
A mgbf mpas-jedi 3dvar has been run as shown below :
![image](https://github.com/user-attachments/assets/c7964803-a3aa-4ca6-b230-e70c38424385)
This demonstrates that the implementation of MGBF with JEDI, which consists of interpolation and its adjoint (using gsi::unstructured_interpolator) as the outer blocks, and MGBF processing on the regular MGBF filtering grids as the central block, works well even for MPAS's unstructured grids.

**Comment by TingLei-NOAA** [comment]
Update on July 29,2024
1) A rather "realistic" mgbf setup using 25km filtering grids (on the same mgbf domain) had been used in regional mpas-jedi  3dvar using one obs and preliminarily showed  reasonable results as below :
![image](https://github.com/user-attachments/assets/6bda62e7-d781-4d6f-886c-6cb8a521cc27)
One current focus is the efficient way to tune/change the characteristic scales of the mgbf based background error covariances.
2) Finished multiple obs /multiple analysis variables mbbf 3dvar for regional fv3-jedi with expected behavior.

**Comment by TingLei-NOAA** [comment]
Ensemble localization using mgbf/saber is working.
The conversion between oops::fields and oops::increment after the localization and its issue when using mgbf/saber have been sorted out and corresponding codes in saber/localization.h have been changed to be able to deal with  the current use of data structures passed from mgbf/saber, which should also work for original nicas localization option. Further evaluation/tuning is ongoing. 
The below is the delp increment from a  Ens3DVar fv3jedi test using 20 members with mgbf (left) and bump::nicas and  using one temperature observation (mgbf localization is using a much larger "effective" cut-off radius)
![image](https://github.com/user-attachments/assets/b026ffd1-7a50-4682-a268-f2b41bf4889c)

**Comment by TingLei-NOAA** [comment]
An update on Aug. 26,2024: 
Currently , a mgbf test used for ensemble localization in the ens3dvar  shows very close results to the one using bump (localization length scales: 240km/0.3) . The vertical profiles of the increment near the  increment center is as below :
![image](https://github.com/user-attachments/assets/6eb5b121-25e4-4997-b614-d051ac032c3e)
The remaining problem is for the mgbf test gives a larger maximum value (17 vs 15).

**Comment by TingLei-NOAA** [comment]
In the ensemble localization, the treatment of 2d variables in mgbf procedure was sorted out and implemented , in which, 2d variables are treated as 3D variables with other vertical levels are initialized as zeros. @MiodragRancic-NOAA  clarified  that this is not as wasteful as at the first glance because those non-zero values spreaded to the upper levels would be also processed in the horizontal directions.
Using one obs test (a single surface observation), it was confirmed this 2dvar treatment capability is working as expected.  One example is the temperature increment profile ( on the horizontally maximum increment center , please note, the length scales in either directions are different for bump and mgbf. The focus is on their qualitative behavior). 
![image](https://github.com/user-attachments/assets/2b08203c-5cdb-4b83-a74c-3875de099229)

**Comment by TingLei-NOAA** [comment]
using the one obtest from GSI and the difference between the analysis and background files (which is not exactly the increment from ensemble covariance and might caused the difference between GSI and other two).   What caused the differences are being investigated
![image](https://github.com/user-attachments/assets/f5243b6f-7490-4b3b-a030-9679be998ee3)

**Comment by TingLei-NOAA** [comment]
Update:  in the previous plot, the GSI results were vertically flipped by a mistake in understanding the vertical order of the increment files from jedi.  Corrected figures showed the problems that fv3-jedi using bump /mgbf created increments of temperature or wind near the top in a pure ens3dvar using surface pressure obs. Further investigation has sorted out  the vertical layout setup (topdown/bottomup) of the fields passed into saber and upgraded mgbf treatment of 2d variables in ensemble localization 
The temperature increment below  showed the right locations of temp increment through covariance between temperature and surface pressure
![image](https://github.com/user-attachments/assets/86179da6-e7c2-4274-9204-94aba42562a6)
,while the differences between gsi/bump/mgbf might be caused by quite a few factors, first the different length scales..

**Comment by TingLei-NOAA** [comment]
Update on Oct. 7,2024:
Two new functions have been added and evaluated recently :  one the normalization step and the mbgf filtering on coarser vertical levels.  Preliminary tests showed they work as expected .

**Comment by TingLei-NOAA** [comment]
update on Oct. 21,2024
1)We have several setup of mgbf to get a dirac response close to 250 km cut-off radius by bump as below
![image](https://github.com/user-attachments/assets/0699e094-24a8-4d51-b3cf-66836d557e20)
2) using of the above setup,  a set of oneob tests (dirac, 3dvar, Ens3dVar and hyb-3dvar) for fv3jedi and (hyb-3dvar) for mpas-jedi had been finished. 
3) continued investing the issue with mgbf_line=.true. and mg_proc=4 in collaboration with MGBF team members at EMC

**Comment by TingLei-NOAA** [comment]
Update on Oct. 28,2024.
A fix in the setup of the mgbf parameters in the current mgbf_lib in jedi allowed mgbf using line filter behave as expected (also avoiding suspicious negative values ) as shown in below: 
![image](https://github.com/user-attachments/assets/6d50d518-ab92-48e5-859b-48a7dbeb8da3)
Further tuning to make the mgbf response, in the horizontal directions,  closer to the gaussian lines corresponding to 250km cut-off radius is ongoing.

**Comment by TingLei-NOAA** [comment]
An satisfying (to some extent, at least ) mgbf setup was found . Its' response from a dirac impulse is shown below : 
![image](https://github.com/user-attachments/assets/043bf510-6164-402a-915e-cf8d697557fd)
If there are not other comments/suggestions from other MGBF team member, I would stick to this setup in the following MGBF tests using real data.

**Comment by TingLei-NOAA** [comment]
The impacts of filtering grid resolution on the smoothness
![image](https://github.com/user-attachments/assets/af8d4c40-5317-4790-8dbc-aa51105015d5)

The model grids are 13km fv3 grids.

**Comment by TingLei-NOAA** [comment]
Update on Dec. 19,2024
Using the 3km conus domain cast from @Junjun-NOAA , a Ens3DVar was run using mgbf for ensemble localization.  The results looks reasonable as shown below (left using bump and right using mgbf)
![image](https://github.com/user-attachments/assets/244699f7-bdd8-4e96-a448-4b2591f077c7)
Currently, mgbf (itself) step used significantly more time than bump , first because bump used less sampling points.  Deep analysis /tuning of mgbf current setup and future upgrading will be discussed in the future.

**Comment by TingLei-NOAA** [comment]
Update on Jan,6,2025
1)Another group of regional MPAS jedi tests (3DVar, ENs3DVar and Hyb) using MGBF for the static B and Ensemble localization. These tests further confirm that MGBF in saber/jedi functions as expected for regional MPAS JEDI, as shown below:
![image](https://github.com/user-attachments/assets/3bfae7de-9989-4ff3-a02a-ffb922ef40a2)
2) current investigation of the computational efficiency of MGBF in saber/jedi was extended to the inside process of MGBF.

**Comment by guoqing-noaa** [comment]
@TingLei-NOAA Great progress! 
I am curious Why the increments are not smooth?

**Comment by TingLei-NOAA** [comment]
@guoqing-noaa  I guess that might be related to the conversion between filtering grids and model grids (more unstructured mpas grids).  As Junjun mentioned in another issue, BUMP recently added a smoother option. We will discuss/decide  what will be done with MGBF/SABER/JEDI.

**Comment by ShunLiu-NOAA** [comment]
@TingLei-NOAA are you using the same conversion method as Masanori for RF?

**Comment by TingLei-NOAA** [comment]
@ShunLiu-NOAA  Yes. The interpolator in SABER::GSIBEC is used by MGBF too!  we are going to use OOPS::interpolator later and wondering if that would help on this smoothness too.

**Comment by TingLei-NOAA** [comment]
The attached is the bar chart for clock times used by various steps inside mgbf when MGBF is used for ensemble localization in a 3km connus domain mpas-jedi run using 144 , 576 and 2304 PEs. 

![Image](https://github.com/user-attachments/assets/6b3aef37-0e3a-4be0-b440-8f893749bc62)

The results showed a rather good computational scalibility thought the imbalances across  PEs on a certain steps are to be investigated .

**Comment by TingLei-NOAA** [comment]
Update on Jan 29,2025: 
@JimPurser-NOAA proposed an approach to calculate the MGBF weights (while using pre-defined aspect tensor, which are much easier to be pre-defined) to mimic pre-defined Gaussian or recursive filter representations of Background error covariances.  Coding, test and evaluation of this approach in 1d situation demonstrate encouraging results . 
An example is as below (the target curve is a gaussian function). 

![Image](https://github.com/user-attachments/assets/5550c768-6999-4383-b9ce-b9e229b72528)

**Comment by TingLei-NOAA** [comment]
Update on Feb. 25,2025
The mgbf in saber/jedi was enhanced with the function to handle inhomogeneous background error properties. The current implementation makes use of the inhomogeneous weights for different grid generations..  In this implementation, accurate normalization in the mapping from filter grids to higher generations is needed when propagating space/generation dependent weights to the right mpi ranks.   Current adjoint interpolation works poorly on this .  M Rancic is working on an upgraded mapping subroutine, which would be used in the current implementation. 
The below is the a comparison between the previous homogeneous results (orange line) (using the "homogeneous"/previous codes) and the new codes (green line)  (inhomogeneous function using homogeneous weights).  Ideally, they should be identical.

![Image](https://github.com/user-attachments/assets/00500e67-0f93-467a-b577-c79fc30f5736)

**Comment by TingLei-NOAA** [comment]
Update on Feb. 26,2025:
The use of oops unstructured interpolator (forward and adjoint) as the outer block for MGBF /SABER had passed the preliminary dirac test for fv3-jedi , as shown below :

![Image](https://github.com/user-attachments/assets/b186dc40-0942-462c-ae5b-c3d31cefc0ec)
The results are almost identical , but, as expected, on the outer block steps, the former only used about 1/3 clock time of the latter. 
When those steps in the former are using the similar clock time as the MGBF as the central block uses, this speeding up means a significant reduction in the clock time for the whole SABER processes.  
This work is still ongoing including its use in association with MGBF for regional mpas jedi.

**Comment by TingLei-NOAA** [comment]
Started using an new adjoint interpolation sub from M. Rancic and investigated and implemented the correct treatment of halo points, the dirac test using inhomogeneous function gave results very close to the results using original homogeneous codes as below
![Image](https://github.com/user-attachments/assets/7adae7b8-8a55-47e8-8878-e7a753748acb)

**Comment by TingLei-NOAA** [comment]
Enhanced interface for MGBF in JEDI to allow the new interpolator  be used as the outer block and improve the accuracy and speed ( ⅓ clock time on those mapping steps). Worked for bother mpas/fv3 jedi.

**Comment by TingLei-NOAA** [comment]
Update (using the mgbf with the tuned vertical filtering parameters)
Working on tuning of mgbf parameters to mimic GSI recursive filter used for RRFSv1 ensemble localization , in which s_sens_v=3, s_ens_h=110km.  According to the conversion , we can first get the horizontal e-folding scale is about 301 km and the same for the vertical direction. The tuned mgbf seems mimic such gaussian curve well as below:
Those test cases are from 13km (horizontal interval of the model grids) connus domain .
![Image](https://github.com/user-attachments/assets/b15ed1f2-0104-4df1-83f1-c569167d0cfb)


The increment for Ens3dvar oneob tests by GSI (Recursive filter) and fv3-jedi (mgbf)

![Image](https://github.com/user-attachments/assets/57d07b50-c1ee-456a-ad6f-4857df955dab)


Regarding the distances (from the center to the bounaries of zeros) , it seems GSI RF has effectively a tiny larger "cut off radius" and a tiny larger value in those tail parts. .  GSI RF also shows a larger value of maximum increment center.  But this is mainly caused by the coarser filtering grid interval of 30km used by MGBF while GSI is using about 13km analysis( and filtering grids). 
If we use about 30km grid intervals for GSI , we have : 
cas
![Image](https://github.com/user-attachments/assets/a5fa16a5-49d0-4210-9d15-0ab55bb9221a)
GSI RF would give a smaller maximum inc . This is understandable, since in this case of GSI, not only filtering grids but the coaser analysis grids would affect hofx and its adjoint part. 
One question to be answered is that when  higher resolution model grids like 3km are used, if the 30km filtering grids would be problematic though the modelled gaussian curve is the same.

**Comment by TingLei-NOAA** [comment]
An update: 
To the previous figure, the Ens3dvar using mgbf used a different setup (smaller lengh scale) compared with the one used in the diract test.  The right one using the same mgbf parameters as in the dirac test shows results below : 

![Image](https://github.com/user-attachments/assets/135fe9bf-7813-4d11-b8cd-c381953361da)
The previous figures would be kept for it shows the impacts of the analysis grids/filtering grids on the maximum inc.

**Comment by TingLei-NOAA** [comment]
Plot of the vertical cross section

![Image](https://github.com/user-attachments/assets/dbe759ec-a503-418c-95e3-d3b09b2a6f6f)

**Comment by JimPurser-NOAA** [comment]
Hi Ting:
They look fairly similar at least.
Thanks,
Jim

**Comment by danholdaway** [comment]
Thank you very much @TingLei-NOAA for getting these side by side single obs tests going. That looks like a fairly decent match between GSI/RF and JEDI/MGBF at first glance. Do we understand the more significant difference in the vertical plots with the larger negative values for MGBF. Is that something we would expect to see with MGBF?

**Comment by TingLei-NOAA** [comment]
Dan
On my mind there are two factors. First as shown in the dirac test  the
mgbf shows a little higher values in the tails ; second. Maybe more
important , what rf really does is an approximation of its theoretical
curve.
I think this might also contribute the difference between rf and mgbf
I would see what Jim and other colleagues think on it.
On my side I would do some further tuning to let mgbf closer to rf
including avoid those obvious negative areas.
Ting


______________________________

Ting Lei

Physical Scientist, Contractor with Lynker in support of

EMC/NCEP/NWS/NOAA

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

***@***.***

301-683-3624

**Comment by TingLei-NOAA** [comment]
@danholdaway   As we discussed in the meeting,   the "negative" values area in the above mgbf  should come from an larger effective length scale modelled by MGBF and a better tuned mgbf would avoid this.  To confirm this, I ran a GSI RF using larger length scale (s_ens_v=5 ( before 3)  and s_ens_h=150km (before 110km).  The results showed similar negative value regions as expected. 

![Image](https://github.com/user-attachments/assets/01a5fc89-867f-44ad-a940-0075680f3090)

**Comment by danholdaway** [comment]
Thanks @TingLei-NOAA, this is really encouraging to see.

**Comment by TingLei-NOAA** [comment]
Update on Apr. 16,2024
First, for clarification, the relationship between the so called e-folding scale and the RF length scale (like s_ens_h, s_ens_v) of GSI recursive filter for ensemble localization is L(e-folding)=L(rf)*sqrt(2) as described in the comment in gsimod.F90. 
Second, using the above relationship, the MGBF setup closely approximating the gaussian curve  in dirac test also approximated reasonably well the GSI RF results when they are used for ensemble localization.  The following figures show one ob test in ens3dvar using GSI (RF ) and FV3-JEDI (MGBF)

![Image](https://github.com/user-attachments/assets/e84d0530-3cd1-4add-a7e3-b7931ddf1b1c)

![Image](https://github.com/user-attachments/assets/0074d8c8-82a8-40bf-9d2f-78b5df12b9f4)

![Image](https://github.com/user-attachments/assets/7b484585-0dbd-4746-a174-54e580c68c66)

![Image](https://github.com/user-attachments/assets/bc10f3cb-912c-412b-9f5f-551e6cda1217)

![Image](https://github.com/user-attachments/assets/768c1ea0-20fe-4eed-8ad3-d2b464707c9c)

So, while there are still some tests/tuning to be done including the behaviors close the boundaries, the major focus would be to real data evaluation and the investigation into the computational efficiencies using MGBF and potential speeding up like usage of reduced resolution in the vertical direciton for MGBF.

**Comment by danholdaway** [comment]
Great stuff @TingLei-NOAA!

**Comment by TingLei-daprediction** [comment]
An real data case is used to compare MGBF and RF in Ens3DVar by fv3-jedi and GSI respectively. In this case, only sonde and rass temperature observation for 2022052619 is used. 
 After quite some efforts to prevent the blurring impacts from other differences betwee fv3-jedi and GSI , the latest results show the differences mainly come the different usages of observations, while how that difference was created is being investigated. 
 An example to show such differences in the increment is as below : 

![Image](https://github.com/user-attachments/assets/ccd4bff8-35da-4372-a538-fa3422be3a39)

**Comment by TingLei-NOAA** [comment]
Thanks to Masanori, direct use of his real case, Rather similar increments from Ens3DVar by GSI using recursive filter and FV3-jedi using MGBF for ensemble localization are observed, for example as below :

![Image](https://github.com/user-attachments/assets/d4eecfb9-f31c-4f0a-8e18-6f886fd8a70d)

**Comment by danholdaway** [comment]
Very nice! Thanks @TingLei-NOAA @Masanori-NOAA

**Comment by TingLei-NOAA** [comment]
For the current 12km conus domain mpas-jedi case, a Ens3DVar using mgbf for the ensemble localization is finished. 

![Image](https://github.com/user-attachments/assets/5fef6233-a9b6-4b04-a93e-9aa2250d3eff)

Using the aircraft wind and temperature observations, the mgbf Ens3DVar (right) demonstrated similar increments compared with the one using BUMP (left):

![Image](https://github.com/user-attachments/assets/90746dfc-8bbc-4bad-8ea6-31a1702de5dd)

Current mgbf setup (for example, no reduction in the vertical filtering resolutons) requires obvious more computational time than bump .
Investigation of the computational efficiencies and enhancements with mgbf would be the focus in the following work

**Comment by TingLei-NOAA** [comment]
A new MGBF setup for Jedi ensemble localization  has been specified, which uses two generations.

![Image](https://github.com/user-attachments/assets/6754ae72-a921-408c-a180-28de1d982302)

Very similar to the previous 4 generation setup, it mimics the GSI recursive filters in rrfsv1 closely while reducing the filtering time by around 10%

**Comment by MiodragRancic-NOAA** [comment]
Ting,

I have a pretty clear idea how to modify the code for the new setup which
will eliminate present two applications of the filter  due to a successive
application of generations one and higher generations.  By then, I think
that you should use generations two and three, not one and two for
localization, which will have the same effect. Also, you should apply all
steps that Sho did and got a remarkable speed up in comparison with
recursive filter.  And finally, we must understand how JEDI maps analysis
grid to first generation of filter grid and either take advantage of that
or build our own code for that.  And finally, we must eliminate clear bugs
in your code that exist even in triad version of the line code that you use
but especially in the fast isotropic version that you should be using as
Sho did.

Miodrag Rancic
Lynker at NOAA/NWS/NCEP/EMC
5830 University Research Ct, Rm 2786
College Park, MD 20740
301-683-3732

**Comment by TingLei-NOAA** [comment]
update on June 17,2025
1) 12 km conus domain 3dEnVar using mgbf had been successfully finished for 23 hourly cycling in one day. Now, working on using D. Lippi's base experiment. 
2) In collaboration with MGBF developers on the improvements of MGBF computational efficiency.   
A reduced vertical resolution (30 in place of previous 59)  set up was tuned and to be used in the cycling experiment, The former mimics very well the previous behavior as below : 

![Image](https://github.com/user-attachments/assets/c9b010b4-c09a-426a-944e-c71627444a53)

3) a couple of important upgrading of the mgbf parallelization have been discussed.  Ting is planning to work one of them :  do the mgbf filtering also in parallel across the vertical levels and variables ( normally, more than 300 vertical levels for all control variables). This significantly enhance the computational scalability while the relatively coarser filtering grids are used.

**Comment by TingLei-NOAA** [comment]
Update on June 23,2025
1) successfully incorporated D. Lippi's setup of his base experiment into the local rrfs workflow branch. The generated experiments using mgbf and bump for localizations had finished one day hourly cycling successfully alongside with the control run using bumploc. Stats for the total clock times and total memory usages of each cycles for the experiments were extracted and analyzed.
The computational time and memory usages were visualized as below 

![Image](https://github.com/user-attachments/assets/36b000d1-ac24-4860-8fc4-2ac9abc12340)
2) 2) mgbf/saber works with only 1 generation applied after fixing an array out-of-bounds index issue in mgbf lib. and verified as it reproduced the identical results as the run of 4 generations but only first G has non-zeror weights. The 1 G setup would reduce significantly the clock time and memory usages but its' capability to mimic the target gaussian distribution and the impact of the differences in the closeness to the target gaussian distributions from 1 G and 2 G setup need more investigations.
3) In collaboration, continued analysis of the mgbf lib regarding its potential optimization, current focus being on the upsending step, is ongoing.

**Comment by hongli-wang** [comment]
> For the current 12km conus domain mpas-jedi case, a Ens3DVar using mgbf for the ensemble localization is finished.
> 
> ![Image](https://github.com/user-attachments/assets/5fef6233-a9b6-4b04-a93e-9aa2250d3eff)
> 
> Using the aircraft wind and temperature observations, the mgbf Ens3DVar (right) demonstrated similar increments compared with the one using BUMP (left):
> 
> ![Image](https://github.com/user-attachments/assets/90746dfc-8bbc-4bad-8ea6-31a1702de5dd)
> 
> Current mgbf setup (for example, no reduction in the vertical filtering resolutons) requires obvious more computational time than bump . Investigation of the computational efficiencies and enhancements with mgbf would be the focus in the following work

Is the extended domain on rotated lat-lon grid like GSI/FV3-LAM?

**Comment by TingLei-NOAA** [comment]
@hongli-wang now, I am just using a lambert projection like 
```
       inner geometry:
           function space: StructuredColumns
           grid:
             type : "regional"
             nx : 224
             ny : 224
             dx : 35.0e3
             dy : 35.0e3
             lonlat(centre) : [261, 40]
             projection :
               type       : lambert_conformal_conic
               latitude0  : 40
               longitude0 : 261
               units: "degrees"
             y_numbering : 1
           partitioner: checkerboard

           halo: 1
```
We might try other grids like lat/lon in the future to more accurately match the mgbf's requirements.

**Comment by TingLei-NOAA** [comment]
Recently,  the MGBF/SBER had been integrated with Jim's algorithm and key subroutine to use vertically stretched filtering grid to mimic vertically varied aspect tensors in the original uniform filtering grids.  The current implementation showed rather encouraging results for relative simple variation pattern like linearly changing aspect tensor profiles.
For, as below, the function specified the vertically stretched grid with  an "uniform" aspect tensor profile to mimic the given aspect tensor profile : 
The red horizontal bars are the calculated aspect tensor on the vertical stretched grids (only 10 vertical levels are "uniformly" (in terms of the indices) selected from the total 30 levels of the stretched grids)

<img width="596" height="364" alt="Image" src="https://github.com/user-attachments/assets/10da9e98-bbe3-462a-979c-f8ef2b430339" />

Second, we selected three dirac tests  on 3 vertical levels with corresponding  aspect tensors in the above profile to compare with the new function with the input aspect tensor profile . The results are very encouraging, while the results (the shape) are very close when different normalization coefficients were applied.  

<img width="1493" height="587" alt="Image" src="https://github.com/user-attachments/assets/722355d0-8aa4-4a90-bd97-2b8419cb53b7" />

<img width="809" height="323" alt="Image" src="https://github.com/user-attachments/assets/94e6b489-dd16-4936-8e07-b2dd339cf3f1" />

<img width="857" height="310" alt="Image" src="https://github.com/user-attachments/assets/b79f54f2-7776-471c-9539-1c58cab0f1e5" />

Its performance with more complicated/challenging aspect tensor profiles (like bigger changing ranges) will be investigated and J. Purser is also considering more robust algorithms to fully meet the possible situation-dependent requirements.

**Comment by TingLei-NOAA** [comment]
@Masanori-NOAA 's 3DEnvar using mgbf for ensemble localization showed the obviously larger maximum increment on the surface, in comparison with GSI.  Investigated this issue and it is found this was caused by the larger dirac response of the mgbf on the surface compared with responses on the inner points.  When a constant normalization coefficient is used this means the mgbf ensemble localization added the ensemble spread and caused larger increments.  This confirmed to results in past evaluation and this issue needs to be addressed when real data applications are being carried out. 
Using the same setup , on the inner vertical level,  the dirac response of MGBF (blue line) is good to mimic the targeted gaussian (GSI RF mimic too, orange line), 

<img width="591" height="229" alt="Image" src="https://github.com/user-attachments/assets/0fb3981d-0c20-4976-afb6-c79d3298e2b3" />

On the surface, it is shown a larger peak was generated. 

<img width="598" height="230" alt="Image" src="https://github.com/user-attachments/assets/b7e44945-b477-4b09-a54f-dd1427a06fb2" />

It is good that the shape distortion is not severe as shown in the renormalized response 

<img width="971" height="368" alt="Image" src="https://github.com/user-attachments/assets/fc93b366-0561-41d5-b817-735befd1a51a" />

Hence, to resolve this issue,  the  existing normalization profile would be filled with vertically varied coefficients to be obtained later rather than the current one constant value.

**Comment by TingLei-NOAA** [comment]
An oops issue was opened for the precision issue when the source grids are on lower resolutions
https://github.com/JCSDA-internal/oops/issues/3012
When some colleagues might have no access to the above link, I pasted it below : 
```
The unstructured oops::interpolator plays an powerful tool for functions in jedi including the current MGBF/SABER uses it (through saber::interpolation) as the outer blocks, which shows obvious improvements in precision and speeding than the older interpolator.
But our work still find that when the source grids (filtering grids used by the central block like MGBF or any other optional block like bump or ID) are on lower resolutions , there would be recognizable position displacement from the interpolation and its adjoint.
two dirac test cases are described below: in which the maximum values of the response should be at the same location where the impulses are put, ideally.
The regional fv3 model grids are on 13km horizontal interval. The source grids (the filtering grids) are 18km or 35km.
The central block is bump::ID
The two yamls could be found here [18km case](https://github.com/TingLei-NOAA/my-debug-files/blob/main/bump_ID-D1-18kmv6-dirac.yaml) and [35km](https://github.com/TingLei-NOAA/my-debug-files/blob/main/bump_ID-D1-35kmv6-dirac.yaml)
The dirac location are at z=29 and X=149 corresponding to the coordinates displayed in the below figure (which are the model grid index starting from zero)
But as shown below, when the filtering grids are 35km, there are one model grid displacement (at x=148 , the maximum value ) . and when the filtering interval is 18km, it is correct (at x=149, the maximum value).

Image First, please see if I missed anything or there are any existing options to improve the "inaccuracy". Second, if such inaccuracy does exist, it is the best performance in terms of precision we could expect now for an unstructured interpolator?
@fmahebert , your comments/suggestions are appreciated.
I also tag EMC colleagues who have been engaged with investigation into the issue
@JimPurser-NOAA @MiodragRancic-NOAA @ManuelPondeca-NOAA
```
The image pointed in the above  is 

<img width="1492" height="588" alt="Image" src="https://github.com/user-attachments/assets/2b9443ad-8c23-4ab7-a1c2-07081022defc" />

**Comment by TingLei-NOAA** [comment]
Update on Apr.28,2026
The impacts from the resolution of the filtering grids compared with the analysis grids in the current MGBF/SABER setup in jedi
@HuiLiu-NOAA reported that in the 13km conus 3dEnVar run using MGBF for ensemble localization, there are noticed "square/rectangle" pattern as shown in the oneob test below

<img width="532" height="655" alt="Image" src="https://github.com/user-attachments/assets/fc123da4-4aca-43b8-9e39-4cf339756cc3" />

Further investigation has identified such an pattern came from the mapping between the analysis grids (13 km horizontal grid spacing) and the  B matrix ( the filtering grid) interface grid spacing (30 km in the horizontal direction)--- the inner grids in the outer block of saber.  When it was changed to 15km as below 
```
     saber outer blocks:
      - saber block name: interpolation
        inner geometry:
          function space: StructuredColumns
          grid:
            type: regional
            nx: 448
            ny: 448
            dx: 15.0e3
            dy: 15.0e3

...
```
But, the mgbf filtering was still done on 30 km grid spacing (so, an extra mapping happened inside mgbf itself between 30km and 15km, howver that is a mapping/filtering between regular grids ).  
The increment as shown below has not such artifacts.   

<img width="519" height="572" alt="Image" src="https://github.com/user-attachments/assets/a0bdb6ea-338a-462f-b228-7bd5b1889c3a" />

Such  artifacts would be less noticeable in real data situations while multiple observations  having overlayed impact regions.  
But it is still a tradeoff between the accuracies and the computational efficiencies.

**Comment by TingLei-NOAA** [comment]
Update on May,26,2026:
Attained a working limited version using Jim Purser’s new codes for mgbf line filter in both horizontal and vertical directions, while the order of the various steps in the filtering are not changed yet. it is encouraging to see more constant magnitudes of the filtering response using the new codes in the vertical direction as shown below :

<img width="1486" height="591" alt="Image" src="https://github.com/user-attachments/assets/de61173d-71c3-4183-9ff4-f489c41774b4" />

<img width="1152" height="450" alt="Image" src="https://github.com/user-attachments/assets/6537b665-a43b-4b09-9424-bde76f4c7e22" />
We see there are differences in the shape from the old and new mgbf codes ("jim-dbz-mgbf" indicates results using the new  codes). Right now, we would regard it as accepted differences in their properties.    The encouraging differences are the more constant magnitudes of filtering responses in the vertical directions . The new codeds gave results 6.01 ( impulse at the 29th vertical level, 0 starting index) and 6.08 (impulse at the 1th vertical level) , obvious improvement in the expected constancy compared with the results from the old codes (0.88 vs  1.51) .  When a different order of the filtering components is used as J. Purser designed for this new scheme, it is supposed the full potential of this new codes will be realized.

**Comment by TingLei-NOAA** [comment]
Update on July,2,2026:
As discussed in July,1,2026 MGBF meeting,  @JimPurser-NOAA  and I agree the currently implemented Jim's recent calibrated MGBF filtering into saber:: mgbf is working as expected.  Namely, on filtering grids, for one generation,  the impulse (of value one) would always get a response of the maximum value of the center of 1.  The remaining question is how to deal with the impacts from the mapping between analysis grids and filtering grids ,while our final goal is the normalization on the analysis grids. However, as discussed below, this function would (if not have) greatly reduce the effort for normalization of MGBF.  
Thanks to Jim for his method and the core mgbf filtering codes from him. 
A few problems/enhancement are being considered. 
1) possible enhancement needed when multiple generations are applied.
2) For computational efficiency, we could also use reduced resolution vertical filtering grids. the mapping between it and the analysis grids is down inside MGBF. Hence, an enhancement to account for this is to be designed/implemented . This would achieve a vertically-constant normalization coefficients. 
3) The mapping between analysis grids and filtering grids in the horizontal directions will be revisited when the targeted generic inner grids (analysis grids) are available.  When the later is present,  easier approaches to account for the mapping from MGBF is possible. 
The table below demonstrate the impacts from the mapping in the horizontal directions (the first 3 rows) and the "accuracy" of the method to get the 1 of the maximum response on the filtering grids (the last 2 rows). 

<img width="1748" height="296" alt="Image" src="https://github.com/user-attachments/assets/e3efc8cc-9a27-451a-af4d-6bf7e3825884" />
